pipeline/file_pulling.go

```diff
@@ -0,0 +1,152 @@
+package pipeline
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+// scheduleFilePulling schedules the file pulling
+func (p *Pipeline) scheduleFilePulling() {
+	// recover from panic
+	// if panic occurs, we will log the error and restart the file pulling
+	defer func() {
+		if r := recover(); r != nil {
+			p.logger.Printf("error occurred when pulling data: %v", r)
+			go p.scheduleFilePulling()
+		}
+	}()
+
+	nextIterationInMs := p.dataFilePullEveryMs
+
+	isFirstRun := true
+
+	for {
+		// if this is the first run, we don't need to wait
+		if isFirstRun {
+			isFirstRun = false
+		} else {
+			<-time.After(time.Duration(nextIterationInMs) * time.Millisecond)
+		}
+
+		p.logger.Printf("Pulling data from %s", p.dataFileUrl)
+
+		fileResponse, err := doDataFileRequest(p.dataFileUrl)
+		if err != nil {
+			p.logger.Printf("failed to pull data file: %v", err)
+			if fileResponse.retryAfter > 0 {
+				p.logger.Printf("received retry-after, retrying after %d seconds", fileResponse.retryAfter)
+
+				// retry after the specified time
+				nextIterationInMs = fileResponse.retryAfter * 1000
+			} else {
+				p.logger.Printf("retrying after 1 second")
+				// retry after 1 second, since we have unhandled error
+				// this can happen from network stutter or something else
+				nextIterationInMs = 1000
+			}
+			continue
+		}
+
+		p.logger.Printf("data file pulled successfully")
+
+		// write to dataFile
+		err = os.WriteFile(p.dataFile, fileResponse.buffer.Bytes(), 0644)
+		if err != nil {
+			p.logger.Printf("failed to write data file: %v", err)
+			// retry after 1 second, since we have unhandled error
+			// this can happen from disk write error or something else
+			nextIterationInMs = 1000
```

Comment:
this should maybe be configurable?

Reply:
if it is not configurable in other implementations, then let's stick to their default value

Reply:
please check the other APIs and resolve the comment